domain/blockdevice/state/state.go

```diff
@@ -0,0 +1,401 @@
+	db, err := st.DB()
+	if err != nil {
+		return "", 0, errors.Trace(err)
+	}
+
+	var (
+		machineUUID string
+		life        domain.Life
+	)
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		var err error
+		machineUUID, life, err = getMachineInfo(ctx, tx, machineId)
+		return errors.Trace(err)
+	})
+	return machineUUID, life, errors.Trace(err)
+}
+
+func getMachineInfo(ctx context.Context, tx *sqlair.TX, machineId string) (string, domain.Life, error) {
+	q := `
+SELECT machine.life_id AS &M.life_id, machine.uuid AS &M.machine_uuid
+FROM   machine
+WHERE  machine.machine_id = $M.machine_id
+`
+	stmt, err := sqlair.Prepare(q, sqlair.M{})
+	if err != nil {
+		return "", 0, errors.Trace(err)
+	}
+
+	result := sqlair.M{}
+	err = tx.Query(ctx, stmt, sqlair.M{"machine_id": machineId}).Get(result)
+	if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+		return "", 0, errors.Trace(err)
+	}
+	if len(result) == 0 {
+		return "", 0, fmt.Errorf("machine %q %w", machineId, errors.NotFound)
+	}
+	life, ok := result["life_id"].(int64)
+	if !ok {
+		return "", 0, errors.Errorf("missing life value for machine %q", machineId)
+	}
+	machineUUID := result["machine_uuid"].(string)
+	return machineUUID, domain.Life(life), nil
+}
+
+// SetMachineBlockDevices sets the block devices visible on the machine.
+// Previously recorded block devices not in the list will be removed.
+func (st *State) SetMachineBlockDevices(ctx context.Context, machineId string, devices ...blockdevice.BlockDevice) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		machineUUID, life, err := getMachineInfo(ctx, tx, machineId)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if life == domain.Dead {
+			return errors.Errorf("cannot update block devices on dead machine %q", machineId)
+		}
+		existing, err := loadBlockDevices(ctx, tx, machineId)
+		if err != nil {
+			return errors.Annotatef(err, "loading block devices for machine %q", machineId)
+		}
+		if !blockDevicesChanged(existing, devices) {
+			return nil
+		}
+
+		if err := updateBlockDevices(ctx, tx, machineUUID, devices...); err != nil {
+			return errors.Annotatef(err, "updating block devices on machine %q (%s)", machineId, machineUUID)
+		}
+		return nil
+	})
+
+	return errors.Trace(err)
+}
+
+func updateBlockDevices(ctx context.Context, tx *sqlair.TX, machineUUID string, devices ...blockdevice.BlockDevice) error {
+	fsTypeQuery := `SELECT * AS &FilesystemType.* FROM filesystem_type`
+	fsTypeStmt, err := sqlair.Prepare(fsTypeQuery, FilesystemType{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	var fsTypes []FilesystemType
+	if err := tx.Query(ctx, fsTypeStmt).GetAll(&fsTypes); err != nil {
+		return errors.Trace(err)
+	}
+	fsTypeByName := make(map[string]int)
+	for _, fsType := range fsTypes {
+		fsTypeByName[fsType.Name] = fsType.ID
+	}
```

Comment:
All of this file system type stuff should go after `if len(devices) == 0`. Why do the hard work if we don't have to.

Reply:
Ah, yeah, I meant to do that